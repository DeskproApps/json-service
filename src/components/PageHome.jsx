import React from 'react';
import PropTypes from 'prop-types';
import Handlebars from 'handlebars';

import { Loader, Content } from '@deskpro/apps-components';

/**
 * Renders the defined template.
 */
class PageHome extends React.Component
{
  static propTypes =
  {
    location: PropTypes.object.isRequired,

    history: PropTypes.object.isRequired,

    dpapp: PropTypes.object.isRequired,
  };


  state = { loaded: false, html: '' };

  /**
   * Invoked immediately before a component is mounted
   */
  componentDidMount()
  {
    const { serviceURL, authorization, template, context } = this.props.location.state;
    const { dpapp } = this.props;

    function html(resp)
    {
      const templateContext = { ...context, resp: resp.body };
      const compiledTemplate = Handlebars.compile(template);
      return compiledTemplate(templateContext);
    }

    if (serviceURL) {

      const serviceTemp = Handlebars.compile(serviceURL);
      const fetchUrl = serviceTemp({...context});

      const headers = {};

      if (authorization) {
        headers['x-proxy-header-authorization'] = authorization;
      }

      const fetchParams = {
        method: 'GET',
        headers
      };
      dpapp.restApi.fetchProxy(fetchUrl, fetchParams)
        .then(html)
        .then(template => {
          this.setState({ html: template, loaded: true })
        })
        .catch(error => {
          if (error instanceof Error) {
            dpapp.ui.showErrorNotification(error)
          } else {
            dpapp.ui.showNotification(error, 'error')
          }
          console.error(error);
        })
    }
  }

  componentDidUpdate()
  {
    const { dpapp } = this.props;
    const { title } = this.props.location.state;
    if (title) {
      dpapp.ui.changeTitle(title);
    }
  }

  /**
   * @returns {XML}
   */
  render()
  {
    if (this.state.loaded) {
      return <Content><div dangerouslySetInnerHTML={{ __html: this.state.html }} /></Content>
    }
    return <div className="dp-text-center"><Loader /></div>;
  }
}

export default PageHome;
