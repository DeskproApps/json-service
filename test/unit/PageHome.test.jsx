import React from 'react';
import renderer from 'react-test-renderer';
import { createAppFromProps, EVENT_WEBAPI_REQUEST_FETCH } from '@deskpro/apps-sdk';
import AppEventEmitter from '@deskpro/apps-sdk/lib/Core/AppEventEmitter';

import { createMemoryHistory as createHistory } from "history";


import PageHome from '../../src/components/PageHome';

it('PageHome renders correctly in the normal state', (done) => {


  const outgoingDispatcher = new AppEventEmitter();
  outgoingDispatcher.on("webapi.request.fetch", (resolve, reject, params) => resolve(
      {
        body: {
          "userId": 1,
            "id": 1,
            "title": "delectus aut autem",
            "completed": false
        }
      }
    )
  );
  const dpapp = createAppFromProps({
    contextProps : {
      type:       'ticket',
      entityId:   '1',
      locationId: 'ticket-sidebar',
      tabId:      'tab-id',
      tabUrl:     'http://127.0.0.1'
    },
    instanceProps : {
      appId:          '1',
      instanceId:     '1',
      appTitle:       'Json Service',
      appPackageName: 'app-boilerplate-react'
    },
    outgoingDispatcher
  });
  const history = createHistory({ initialEntries: ["home"], initialIndex: 0});

  const state = {
    title:     "The title displayed in the toolbar",
    template:   "<h1>{{ resp.title }}</h1>\n" +
    "<p>Agent: {{ tab.agent.first_name }}</p>\n" +
    "<p>UserId: {{ resp.userId }}</p>",
    serviceURL: "https://jsonplaceholder.typicode.com/todos/1",
    context : {
      tab: {
        agent: {
          first_name: test
        }
      },
      me : {
        id: 100
      }
    }
  };
  const tree = renderer.create(<PageHome
      dpapp=      {dpapp}
      history=    {history}
      location=   {{ state }}
    />
  );

  setTimeout(function() {
    expect(tree.toJSON()).toMatchSnapshot();
    done()
  }, 500);



});
