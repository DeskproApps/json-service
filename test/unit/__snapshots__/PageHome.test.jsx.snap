// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageHome renders correctly in the normal state 1`] = `
<div
  className="dp-Content"
>
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": "<h1>delectus aut autem</h1>
      <p>Agent: [object Object]</p>
      <p>UserId: 1</p>",
      }
    }
  />
</div>
`;
